resources/2024-06-resource-extensibility-feature-spec.md

```diff
@@ -0,0 +1,452 @@
+1. Documentation to guide the users on how to test and deploy the custom resource types in Radius
+
+*User groups and their needs*
+
+There are two types of users that we need to cater to:
+1. Users who have registered schema using typespec schema format
+1. Users who have registered schema using yaml schema format
+
+*User experience*
+
+1. Deb compiles the typespec schema 
+    
+        ```bash
+        tsp compile
+        ```
+        The typespec schema is compiled and validated for any errors and warnings
+
+1. Deb publishes the bicep types to a registry
+    
+        ```bash
+        rad publish-bicep-types --file contoso.tsp --registry myregistry
+        ```
+    The `contoso.tsp` is complied into open api specs; the  index.json and types.json files are generated and published to the registry for the custom resource type Plaid and the new resource provider is added to `bicepconfig.json`
+    ```json
+    {
+        "experimentalFeaturesEnabled": {
+            "extensibility": true,
+            "extensionRegistry": true,
+            "dynamicTypeLoading": true
+        },
+        "extensions": {
+            "radius": "br:biceptypes.azurecr.io/radius:latest",
+            "aws": "br:biceptypes.azurecr.io/aws:latest"
+            "Contoso": "br:biceptypes.azurecr.io/contoso:latest"
+        }
+    }
+    ```
+
+1. Create a app.bicep with plaid resource type
+
+    ```bicep
+    param plaidName string
+    param plaidDescription string
+    param plaidQueue string
+
+    resource plaid 'Contoso.Messaging/Plaid@v1' = {
+        name: plaidName
+        description: plaidDescription
+        queue: plaidQueue
+    }
+    ```
+1. Deb uses the rad CLI to deploy the application with the custom resource type Plaid
+
+    ```bash
+    rad deploy app.bicep 
+    ```
+    The application is deployed with the custom resource type Plaid in Radius. Deb can iterate on the changes, compile and test the application with the custom resource type.
+
+#### Ship the custom platform
+
+As a platform engineer, I want to register the API's to the platform, provide documentation, onboarding and support. Intuitive, seamless, collaborative
+
+**Solution in Radius**
+
+1. End-end-documentation for the custom resource types - Radius provides documentation to the users on how to author, test and deploy the custom resource types in Radius.
+    1. Typespec samples and documentation - Provide samples and documentation for the users to author the schema in typespec format
+    1. Documentation for generating Radius-Bicep types - Reuse existing pipelines and GH workflows used in Radius to generate Radius types in Bicep as an example and provide documentation to the users on how to build and publish the custom resource types in Radius.
+
+Question - Documentation seems like a good starting point for users to get started on using custom resource types? What are other ways, should we provide sdk's to author the schema and generate the types?
```

Comment:
I think documentation and samples is the default starting answer and required for anything we build. It also feels like we're already signing up to build some pretty rich tooling (shown in the design here). Let's discuss if you think we should do more.

Reply:
template repositoryry, scaffolding template, tooling